Jackson unable to deserialize commandEndpoint

While testing Hono2.0.0RC3 we encountered an issue while sending commands via the lora adapter. The lora adapter uses command endpoints to identify where to send the commands to. However, this doesn't work since jackson is unable to deserialize the command endpoint, following exception is thrown:
```
2022-06-09 18:13:18,310 WARN  [org.ecl.hon.tra.TracingHelper] (vert.x-eventloop-thread-1) An unexpected error occurred! [logged on trace 9365415a5c9f4bb0aa352a0c67c552b8, span 7f82f70fdf4ea076]: java.lang.IllegalArgumentException: Cannot construct instance of `org.eclipse.hono.util.CommandEndpoint` (no Creators, like default constructor, exist): cannot deserialize from Object value (no delegate- or property-based Creator)
 at [Source: UNKNOWN; byte offset: #UNKNOWN] (through reference chain: org.eclipse.hono.util.RegistrationAssertion["command-endpoint"])
        at com.fasterxml.jackson.databind.ObjectMapper._convert(ObjectMapper.java:4393)
        at com.fasterxml.jackson.databind.ObjectMapper.convertValue(ObjectMapper.java:4324)
        at io.quarkus.vertx.runtime.jackson.QuarkusJacksonJsonCodec.fromValue(QuarkusJacksonJsonCodec.java:86)
        at io.vertx.core.json.JsonObject.mapTo(JsonObject.java:117)
        at org.eclipse.hono.client.registry.amqp.ProtonBasedDeviceRegistrationClient.lambda$assertRegistration$13(ProtonBasedDeviceRegistrationClient.java:203)
        at io.vertx.core.impl.future.Mapping.onSuccess(Mapping.java:35)
        at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60)
        at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211)
        at io.vertx.core.impl.future.Composition$1.onSuccess(Composition.java:62)
        at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60)
        at io.vertx.core.impl.future.SucceededFuture.addListener(SucceededFuture.java:88)
        at io.vertx.core.impl.future.Composition.onSuccess(Composition.java:43)
        at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60)
        at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211)
        at io.vertx.core.impl.future.Composition$1.onSuccess(Composition.java:62)
        at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60)
        at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211)
        at io.vertx.core.impl.future.Mapping.onSuccess(Mapping.java:40)
        at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60)
        at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211)
        at io.vertx.core.impl.future.Composition$1.onSuccess(Composition.java:62)
        at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60)
        at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211)
        at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23)
        at io.vertx.core.impl.future.PromiseImpl.onSuccess(PromiseImpl.java:49)
        at io.vertx.core.impl.future.PromiseImpl.handle(PromiseImpl.java:41)
        at org.eclipse.hono.client.amqp.RequestResponseClient.handleResponse(RequestResponseClient.java:368)
        at org.eclipse.hono.client.amqp.connection.impl.HonoConnectionImpl.lambda$createReceiver$15(HonoConnectionImpl.java:726)
        at io.vertx.proton.impl.ProtonReceiverImpl.onDelivery(ProtonReceiverImpl.java:236)
        at io.vertx.proton.impl.ProtonTransport.handleSocketBuffer(ProtonTransport.java:168)
        at io.vertx.core.net.impl.NetSocketImpl.lambda$new$1(NetSocketImpl.java:98)
        at io.vertx.core.streams.impl.InboundBuffer.handleEvent(InboundBuffer.java:240)
        at io.vertx.core.streams.impl.InboundBuffer.write(InboundBuffer.java:130)
        at io.vertx.core.net.impl.NetSocketImpl$DataMessageHandler.handle(NetSocketImpl.java:418)
        at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:50)
        at io.vertx.core.impl.ContextImpl.emit(ContextImpl.java:274)
        at io.vertx.core.impl.EventLoopContext.emit(EventLoopContext.java:22)
        at io.vertx.core.net.impl.NetSocketImpl.handleMessage(NetSocketImpl.java:394)
        at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:155)
        at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1372)
        at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1235)
        at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1284)
        at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:510)
        at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:449)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:279)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:722)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:658)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:584)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:496)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:833)
        at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:597)
        at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:194)
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `org.eclipse.hono.util.CommandEndpoint` (no Creators, like default constructor, exist): cannot deserialize from Object value (no delegate- or property-based Creator)
 at [Source: UNKNOWN; byte offset: #UNKNOWN] (through reference chain: org.eclipse.hono.util.RegistrationAssertion["command-endpoint"])
        at com.fasterxml.jackson.databind.DeserializationContext.reportBadDefinition(DeserializationContext.java:1904)
        at com.fasterxml.jackson.databind.DatabindContext.reportBadDefinition(DatabindContext.java:400)
        at com.fasterxml.jackson.databind.DeserializationContext.handleMissingInstantiator(DeserializationContext.java:1349)
        at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1415)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:351)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:184)
        at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:277)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:462)
        at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1405)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:351)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:184)
        at com.fasterxml.jackson.databind.ObjectMapper._convert(ObjectMapper.java:4388)
        ... 70 more
```

--- a/core/src/main/java/org/eclipse/hono/util/CommandEndpoint.java
+++ b/core/src/main/java/org/eclipse/hono/util/CommandEndpoint.java
@@ -22,13 +22,19 @@ import java.util.Objects;
 import java.util.Optional;
 
 import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonInclude.Include;
 import com.fasterxml.jackson.annotation.JsonProperty;
+
+import io.quarkus.runtime.annotations.RegisterForReflection;
 
 /**
  * Encapsulates the command endpoint information for a device as defined by the
  * <a href="https://www.eclipse.org/hono/docs/api/device-registration/">Device Registration API</a>.
  */
-public class CommandEndpoint {
+@RegisterForReflection
+@JsonInclude(value = Include.NON_NULL)
+public final class CommandEndpoint {
     private String uri;
     private Map<String, String> headers = new HashMap<>();
     private Map<String, Object> payloadProperties = new HashMap<>();
@@ -64,6 +70,7 @@ public class CommandEndpoint {
      * @return An unmodifiable view on the headers.
      */
     @JsonProperty(value = RegistrationConstants.FIELD_COMMAND_ENDPOINT_HEADERS)
+    @JsonInclude(value = Include.NON_EMPTY)
     public Map<String, String> getHeaders() {
         return Collections.unmodifiableMap(headers);
     }
@@ -86,6 +93,7 @@ public class CommandEndpoint {
      * @return An unmodifiable view on the payload properties.
      */
     @JsonProperty(value = RegistrationConstants.FIELD_COMMAND_ENDPOINT_PAYLOAD_PROPERTIES)
+    @JsonInclude(value = Include.NON_EMPTY)
     public Map<String, Object> getPayloadProperties() {
         return Collections.unmodifiableMap(payloadProperties);
     }

--- a/services/device-registry-base/src/main/java/org/eclipse/hono/service/management/device/Device.java
+++ b/services/device-registry-base/src/main/java/org/eclipse/hono/service/management/device/Device.java
@@ -383,4 +383,25 @@ public class Device {
     public final Set<String> getAuthorities() {
         return Collections.unmodifiableSet(authorities);
     }
+
+    /**
+     * Gets the definition of the endpoint that commands for this device should be sent to.
+     *
+     * @return The endpoint definition or {@code null} if not set.
+     */
+    public final CommandEndpoint getCommandEndpoint() {
+        return commandEndpoint;
+    }
+
+    /**
+     * Sets the definition of the endpoint that commands for this device should be sent to.
+     *
+     * @param endpoint The endpoint definition.
+     * @return A reference to this for fluent use.
+     * @throws NullPointerException if endpoint is {@code null}.
+     */
+    public final Device setCommandEndpoint(final CommandEndpoint endpoint) {
+        this.commandEndpoint = Objects.requireNonNull(endpoint);
+        return this;
+    }
 }

--- a/site/homepage/content/release-notes.md
+++ b/site/homepage/content/release-notes.md
@@ -46,6 +46,8 @@ description = "Information about changes in recent Hono releases. Includes new f
 * The mechanism to delete obsolete `hono.command_internal.*` Kafka topics could have deleted still used topics in
   case the Kubernetes API server gave information about the running containers with a delay of several seconds. This has
   been fixed.
+* The native executable based images of protocol adapters failed to deserialize a registration assertion if it included command
+  endpoint information that had been configured for the device. This has been fixed.
 
 ### API Changes
 

--- a/tests/src/test/java/org/eclipse/hono/tests/registry/DeviceRegistrationApiTests.java
+++ b/tests/src/test/java/org/eclipse/hono/tests/registry/DeviceRegistrationApiTests.java
@@ -13,15 +13,19 @@
 
 package org.eclipse.hono.tests.registry;
 
+import static org.junit.jupiter.api.Assertions.assertAll;
+
 import static com.google.common.truth.Truth.assertThat;
 
 import java.net.HttpURLConnection;
 import java.util.Collections;
 import java.util.List;
+import java.util.Map;
 import java.util.concurrent.TimeUnit;
 
 import org.eclipse.hono.client.registry.DeviceRegistrationClient;
 import org.eclipse.hono.service.management.device.Device;
+import org.eclipse.hono.util.CommandEndpoint;
 import org.eclipse.hono.util.MessageHelper;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
@@ -77,20 +81,31 @@ abstract class DeviceRegistrationApiTests extends DeviceRegistryTestBase {
 
         final JsonObject defaults = new JsonObject()
                 .put(MessageHelper.SYS_PROPERTY_CONTENT_TYPE, "application/vnd.acme+json");
+        final Map<String, String> headers = Map.of("key1", "value1");
+        final Map<String, String> props = Map.of("key2", "value2");
+        final var endpoint = new CommandEndpoint()
+                .setUri("http://some.domain.com")
+                .setHeaders(headers)
+                .setPayloadProperties(props);
         final Device device = new Device();
         device.setDefaults(defaults.getMap());
+        device.setCommandEndpoint(endpoint);
         final String deviceId = getHelper().getRandomDeviceId(tenantId);
 
         getHelper().registry
-                .registerDevice(tenantId, deviceId, device)
-                .compose(r -> getClient().assertRegistration(tenantId, deviceId, null, NoopSpan.INSTANCE.context()))
-                .onComplete(ctx.succeeding(resp -> {
-                    ctx.verify(() -> {
-                        assertThat(resp.getDeviceId()).isEqualTo(deviceId);
-                        assertThat(resp.getDefaults()).containsExactlyEntriesIn(defaults.getMap());
-                    });
-                    ctx.completeNow();
-                }));
+            .registerDevice(tenantId, deviceId, device)
+            .compose(r -> getClient().assertRegistration(tenantId, deviceId, null, NoopSpan.INSTANCE.context()))
+            .onComplete(ctx.succeeding(assertion -> {
+                ctx.verify(() -> {
+                    assertAll(
+                        () -> assertThat(assertion.getDeviceId()).isEqualTo(deviceId),
+                        () -> assertThat(assertion.getDefaults()).containsExactlyEntriesIn(defaults.getMap()),
+                        () -> assertThat(assertion.getCommandEndpoint().getUri()).isEqualTo("http://some.domain.com"),
+                        () -> assertThat(assertion.getCommandEndpoint().getHeaders()).containsExactlyEntriesIn(headers),
+                        () -> assertThat(assertion.getCommandEndpoint().getPayloadProperties()).containsExactlyEntriesIn(props));
+                });
+                ctx.completeNow();
+            }));
     }
 
     /**

